src/librustc_mir/build/expr/stmt.rs

```diff
@@ -79,14 +79,31 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                 block.unit()
             }
             ExprKind::Continue { label } => {
-                this.break_or_continue(expr_span, label, block,
-                                       |loop_scope| loop_scope.continue_block)
+                let LoopScope { continue_block, extent, .. } =
+                    *this.find_loop_scope(expr_span, label);
+                this.break_or_continue(expr_span, block, continue_block, extent)
             }
-            ExprKind::Break { label } => {
-                this.break_or_continue(expr_span, label, block, |loop_scope| {
-                    loop_scope.might_break = true;
-                    loop_scope.break_block
-                })
+            ExprKind::Break { label, mut value } => {
+                let (break_block, extent, destination) = {
+                    let LoopScope {
+                        ref mut might_break,
+                        break_block,
+                        extent,
+                        ref destination,
+                        ..
+                    } = *this.find_loop_scope(expr_span, label);
+                    *might_break = true;
+                    // FIXME unnecessary clone
+                    (break_block, extent, destination.clone())
+                };
+                if let Some(dest) = destination {
+                    if let Some(value) = value.take() {
```

Comment:
```
error[E0499]: cannot borrow `*this` as mutable more than once at a time
  --> src/librustc_mir/build/expr/stmt.rs:91:45
   |
87 |                     let scope = this.find_loop_scope(expr_span, label);
   |                                 ---- first mutable borrow occurs here
...
91 |                             unpack!(block = this.into(dest, block, value))
   |                             ----------------^^^^--------------------------
   |                             |               |
   |                             |               second mutable borrow occurs here
   |                             in this macro invocation
...
97 |                 };
   |                 - first borrow ends here
```

I don't think we can avoid cloning `destination` without unsafe code because the borrow checker doesn't know that `this.into` will never pop the loop scope (among other reasons).